app/models/rules_of_origin/scheme_set.rb

```diff
@@ -0,0 +1,65 @@
+module RulesOfOrigin
+  class SchemeSet
+    class_attribute :current
+    attr_reader :base_path, :links
+
+    def initialize(source_file)
+      data = JSON.parse read_file(source_file)
```

Comment:
How do you feel about not having side effects in your constructor? I guess you could either a) implement a builder for your class `def self.build` or b) pass the reading, parsing and error handled thing to the initialize method.

```ruby
def self.build(source_file)
  data = read_file(source_file)
  data = JSON.parse(data)
  links = data['links'].map(&Link.method(:new_with_check)).compact.freeze
  schemes = data['schemes'].map(&Scheme.method(:new))
                                 .index_by(&:scheme_code)
                                 .freeze
  countries = rebuild_countries_to_schemes_index.freeze
  new(links, schemes, countries)
end
```

I don't think you need the error handling around the ScopeDoesNotMatch since you're in control of the files that are being loaded?

Reply:
Also, do we need to worry about runtime mutation of the objects? Probably not worth marking the containing data structures as immutable? It doesn't recurse through to each item in the containing structure AFAIK, too

Reply:
initialize stuff - yeh, agree - I've actually already got a similar change in a later PR I'm working on - I'll move it onto this. 

The immutable stuff helps with memory usage. I think I get the same benefits using frozen string literals pragma so I'll just add that to the top of all the files.

ScopesDoesNotMatch bit can be dropped - I figured the info was in the file so I'd check it against the version of the site it was getting loaded into - more as a protection against developer confusion.